File: examples/quickstart.ts
// Tape quickstart — TypeScript. The same scenario as the Python / Go / Java siblings.
//
//   node --experimental-strip-types --no-warnings examples/quickstart.ts

import { TapeClient, EffectStatus } from '../tape/sdk/typescript/src/index.ts';

const LANG = 'typescript';
const URL  = process.env.TAPE_URL ?? 'tape://127.0.0.1:7878';

async function main(): Promise<number> {
  console.log(`[quickstart/${LANG}] connecting to ${URL}`);
  const c = new TapeClient(URL);
  try {
    const invocation = `qs-${LANG}-${Date.now()}`;
    const run: any = await c.beginRun({
      appName: 'quickstart', userId: 'quickstart-user',
      sessionId: invocation, invocationId: invocation,
      leaseOwner: `qs-${LANG}`, leaseTtlMs: 60_000,
    });
    console.log(`[quickstart/${LANG}] begin_run    → run-id=${run.runId}`);

    await c.recordDecision({
      runId: run.runId, decisionIndex: 0,
      model: 'quickstart', requestJson: '{}', responseJson: '{}',
    });
    console.log(`[quickstart/${LANG}] record_decision  decision_index=0`);

    const be: any = await c.beginEffect({
      runId: run.runId, decisionIndex: 0,
      toolName: 'hello', callIndex: 0,
      requestJson: JSON.stringify({ who: LANG }),
    });
    console.log(`[quickstart/${LANG}] begin_effect   → key=${be.idempotencyKey}  status=${be.status}`);

    await c.completeEffect({
      runId: run.runId, idempotencyKey: be.idempotencyKey,
      status: EffectStatus.CONFIRMED,
      responseJson: JSON.stringify({ ok: true, who: LANG }),
    });
    console.log(`[quickstart/${LANG}] complete_effect → status=CONFIRMED`);

    const eff: any = await c.getEffect({ runId: run.runId, idempotencyKey: be.idempotencyKey });
    console.log(`[quickstart/${LANG}] get_effect     status=${eff.effect.status}  response=${eff.effect.responseJson}`);
  } finally {
    c.close();
  }
  return 0;
}

main().then(
  (rc) => process.exit(rc),
  (ex) => { console.error(`quickstart/${LANG}: ${ex?.message ?? ex}`); process.exit(1); },
);

File: examples/quickstart.go
// Tape quickstart — Go. The same scenario as the Python / TS / Java siblings.
//
//   go run examples/quickstart.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"time"

	tape "github.com/vamsiramakrishnan/durable-agents/tape/sdk/go"
)

const lang = "go"

func main() {
	url := os.Getenv("TAPE_URL")
	if url == "" {
		url = "tape://127.0.0.1:7878"
	}
	fmt.Printf("[quickstart/%s] connecting to %s\n", lang, url)

	c, err := tape.Dial(url)
	if err != nil {
		fmt.Fprintf(os.Stderr, "dial %s: %v\n", url, err)
		os.Exit(1)
	}
	defer c.Close()

	ctx := context.Background()
	invocation := fmt.Sprintf("qs-%s-%d", lang, time.Now().Unix())

	run, err := c.BeginRun(ctx, tape.BeginRunOpts{
		AppName: "quickstart", UserID: "quickstart-user",
		SessionID: invocation, InvocationID: invocation,
		LeaseOwner: "qs-" + lang, LeaseTTLMs: 60_000,
	})
	check(err, "begin_run")
	fmt.Printf("[quickstart/%s] begin_run    → run-id=%s\n", lang, run.RunId)

	_, err = c.RecordDecision(ctx, run.RunId, 0, "quickstart", "{}", "{}", "", "")
	check(err, "record_decision")
	fmt.Printf("[quickstart/%s] record_decision  decision_index=0\n", lang)

	reqJSON, _ := json.Marshal(map[string]string{"who": lang})
	be, err := c.BeginEffect(ctx, tape.BeginEffectOpts{
		RunID: run.RunId, DecisionIndex: 0,
		ToolName: "hello", CallIndex: 0,
		RequestJSON: string(reqJSON),
	})
	check(err, "begin_effect")
	fmt.Printf("[quickstart/%s] begin_effect   → key=%s  status=%v\n", lang, be.IdempotencyKey, be.Status)

	respJSON, _ := json.Marshal(map[string]any{"ok": true, "who": lang})
	_, err = c.CompleteEffect(ctx, run.RunId, be.IdempotencyKey,
		tape.EffectStatusConfirmed, string(respJSON), "")
	check(err, "complete_effect")
	fmt.Printf("[quickstart/%s] complete_effect → status=CONFIRMED\n", lang)

	got, err := c.GetEffect(ctx, run.RunId, be.IdempotencyKey)
	check(err, "get_effect")
	fmt.Printf("[quickstart/%s] get_effect     status=%v  response=%s\n",
		lang, got.Effect.Status, got.Effect.ResponseJson)
}

func check(err error, where string) {
	if err != nil {
		fmt.Fprintf(os.Stderr, "%s: %v\n", where, err)
		os.Exit(1)
	}
}
